File: frontend/src/types/test-run.ts
export type TestRunStatus = "queued" | "running" | "passed" | "failed" | "error";
export type StepResultStatus = "passed" | "failed" | "skipped" | "error";

export interface StepResult {
  step_number: number;
  status: StepResultStatus;
  action: string;
  expected: string;
  actual: string;
  error_message: string;
  evidence_url: string;
  generated_code: string;
  duration_ms: number;
}

export interface TestRun {
  id: string;
  org_id: string;
  test_case_id: string;
  test_case_title: string;
  test_type: string;
  project_name: string;
  batch_name: string;
  batch_id: string;
  status: TestRunStatus;
  triggered_by: string;
  celery_task_id: string;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number;
  total_steps: number;
  passed_steps: number;
  failed_steps: number;
  error_message: string;
  run_until_step: number;
  intent_overrides: Record<string, string>;
  current_step: number | null;
  results: StepResult[];
  created_at: string;
  updated_at: string;
}

export interface TestRunListResponse {
  items: TestRun[];
  total: number;
  page: number;
  page_size: number;
}

export interface TestRunBatch {
  batch_id: string;
  project_name: string;
  batch_name: string;
  runs: TestRun[];
  total_runs: number;
  passed: number;
  failed: number;
  queued: number;
  running: number;
  error: number;
  status: string;
  created_at: string;
}

export interface BatchListResponse {
  items: TestRunBatch[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/app/models/test_run.py
from datetime import datetime, timezone
from enum import StrEnum

from bson import ObjectId
from pydantic import BaseModel, Field


class TestRunStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PASSED = "passed"
    FAILED = "failed"
    ERROR = "error"


class StepResultStatus(StrEnum):
    PASSED = "passed"
    FAILED = "failed"
    SKIPPED = "skipped"
    ERROR = "error"


class StepResult(BaseModel):
    step_number: int
    status: StepResultStatus = StepResultStatus.SKIPPED
    action: str = ""
    expected: str = ""
    actual: str = ""
    error_message: str = ""
    evidence_url: str = ""
    generated_code: str = ""
    duration_ms: int = 0
    verification_mode: str = ""  # "ai", "dom-only", or "skipped"
    retry_count: int = 0
    ai_heal_attempts: int = 0


class TestRun(BaseModel):
    id: ObjectId | None = Field(default=None, alias="_id")
    org_id: ObjectId
    test_case_id: ObjectId
    test_case_title: str
    test_type: str = "ui"
    status: TestRunStatus = TestRunStatus.QUEUED
    triggered_by: ObjectId
    project_name: str = ""
    batch_name: str = ""
    batch_id: str = ""
    celery_task_id: str = ""
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int = 0
    total_steps: int = 0
    passed_steps: int = 0
    failed_steps: int = 0
    error_message: str = ""
    run_until_step: int = 0  # 0 = run all steps; N = run steps 1..N only
    intent_overrides: dict[str, str] = Field(default_factory=dict)  # step_number (str) → intent JSON
    results: list[StepResult] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    model_config = {
        "populate_by_name": True,
        "arbitrary_types_allowed": True,
        "json_encoders": {ObjectId: str},
    }
